### PR TITLE
[VALIDATED_STATE] - Give proposal and validation functions access to parent leaf

### PR DESCRIPTION
**Pull request by shenkeyao**
Closes https://github.com/EspressoSystems/HotShot/issues/2662.

### This PR: 
* Replaces the `parent_header` argument with `parent_leaf` in `Header::new` and `validate_and_apply_header`.
* Adds `NodeType` generic type to the block header and validated state traits and testing structs.

### Key places to review: 
* `crates/types/src/traits/block_contents.rs`.
* `crates/types/src/traits/states.rs`.

### CODE REVIEW

File: crates/example-types/src/block_types.rs [hunk 1]
```diff
@@ -180,25 +180,25 @@ pub struct TestBlockHeader {
     pub payload_commitment: VidCommitment,
 }
 
-impl BlockHeader for TestBlockHeader {
+impl<TYPES: NodeType> BlockHeader<TYPES> for TestBlockHeader {
     type Payload = TestBlockPayload;
     type State = TestValidatedState;
```
Comment (jbearer) [inline crates/example-types/src/block_types.rs:185]: I wonder if we can remove either of these associated types and refer to them as `TYPES::Something` now

File: crates/task-impls/src/consensus.rs [hunk 2]
```diff
@@ -618,7 +620,7 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, A: ConsensusApi<TYPES, I> +
                 };
                 let Ok(state) = parent_state.validate_and_apply_header(
                     &consensus.instance_state,
-                    &parent_leaf.block_header.clone(),
+                    &parent_leaf.clone(),
```
Comment (jbearer) [inline crates/task-impls/src/consensus.rs:623]: Why are we cloning this if we're passing it in by reference?

**Review by jbearer — APPROVED** [review]

File: crates/example-types/src/block_types.rs (hunk 1 above)
Comment (shenkeyao) [reply crates/example-types/src/block_types.rs:185]: I think we can! Let me try this.

File: crates/task-impls/src/consensus.rs (hunk 2 above)
Comment (shenkeyao) [reply crates/task-impls/src/consensus.rs:623]: I don't the cloning is necessary. I'll try removing it.

File: crates/example-types/src/block_types.rs (hunk 1 above)
Comment (shenkeyao) [reply crates/example-types/src/block_types.rs:185]: Also removed these types from `ValidatedState`.

**Review by jbearer — APPROVED** [review]

**Pull request merged**